assets/app/scripts/controllers/deployments.js

```diff
@@ -85,6 +120,222 @@ angular.module('openshiftConsole')
       }
     }
 
+    $scope.startLatestDeployment = function(deploymentConfigName) {
+      var deploymentConfig = $scope.deploymentConfigs[deploymentConfigName];
+
+      // increase latest version by one so starts new deployment based on latest
+      var req = {
+        kind: "DeploymentConfig",
+        apiVersion: "v1",
+        metadata: deploymentConfig.metadata,
+        spec: deploymentConfig.spec,
+        status: deploymentConfig.status
+      };
+      if (!req.status.latestVersion) {
+        req.status.latestVersion = 0;
+      }
+      req.status.latestVersion++;
+
+      // update the deployment config
+      DataService.update("deploymentconfigs", deploymentConfigName, req, $scope).then(
+        function() {
+            $scope.alerts = [
+            {
+              type: "success",
+              message: "Deployment #" + req.status.latestVersion + " of " + deploymentConfigName + " has started.",
+            }
+          ];
+        },
+        function(result) {
+          $scope.alerts = [
+            {
+              type: "error",
+              message: "An error occurred while starting the deployment.",
+              details: getErrorDetails(result)
+            }
+          ];
+        }
+      );
+    };
+
+    $scope.retryFailedDeployment = function(deploymentConfigName, deploymentName) {
+      var deployment = $scope.deploymentsByDeploymentConfig[deploymentConfigName][deploymentName];
+      var req = deployment;
+
+      // delete the deployer pod as well as the deployment hooks pods, if any
+      DataService.list("pods", $scope, function(list) {
+        var pods = list.by("metadata.name");
+        var deleteDeployerPod = function(pod) {
+          var deployerPodForAnnotation = $filter('annotationName')('deployerPodFor');
+          if (pod.metadata.labels[deployerPodForAnnotation] === deploymentName) {
+            DataService.delete("pods", pod.metadata.name, $scope).then(
+              function() {
+                Logger.info("Deployer pod " + pod.metadata.name + " deleted");
+              },
+              function(result) {
+                $scope.alerts = [
+                  {
+                    type: "error",
+                    message: "An error occurred while deleting the deployer pod.",
```

Comment:
Using the selector would be better at least, then we arent getting ALL the
pods.  Is there a timeframe for the API endpoints?

On Wed, Aug 26, 2015 at 10:26 AM, Dan Mace <notifications@github.com> wrote:

> In assets/app/scripts/controllers/deployments.js
> <https://github.com/openshift/origin/pull/4177#discussion_r37987743>:
>
> > +
> > +      // delete the deployer pod as well as the deployment hooks pods, if any
> > +      DataService.list("pods", $scope, function(list) {
> > +        var pods = list.by("metadata.name");
> > +        var deleteDeployerPod = function(pod) {
> > +          var deployerPodForAnnotation = $filter('annotationName')('deployerPodFor');
> > +          if (pod.metadata.labels[deployerPodForAnnotation] === deploymentName) {
> > +            DataService.delete("pods", pod.metadata.name, $scope).then(
> > +              function() {
> > +                Logger.info("Deployer pod " + pod.metadata.name + " deleted");
> > +              },
> > +              function(result) {
> > +                $scope.alerts = [
> > +                  {
> > +                    type: "error",
> > +                    message: "An error occurred while deleting the deployer pod.",
>
> Unfortunately, you do actually need to clean up deployers before resetting
> states for a retry (duplicating logic currently all bound up in the deploy
> CLI command). We have a cart to introduce API endpoints for
> deploy/retry/cancel in this trello card <https://trello.com/c/qCzTBO0h>.
>
> You can select all deployer pods using a label selector:
>
> openshift.io/deployer-pod-for.name=<deployment-name>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4177/files#r37987743>.
>